crates/optimism/chainspec/src/lib.rs

```diff
@@ -178,6 +188,58 @@ pub struct OpChainSpec {
     pub inner: ChainSpec,
 }
 
+impl OpChainSpec {
+    /// Read from parent to determine the base fee for the next block
+    pub fn next_block_base_fee(&self, parent: &Header, timestamp: u64) -> U256 {
+        let is_holocene = self.inner.is_fork_active_at_timestamp(
+            reth_optimism_forks::OptimismHardfork::Holocene,
+            timestamp,
+        );
+        // If we are in the Holocene, we need to use the base fee params from the parent block's
+        // nonce Else, use the base fee params from chainspec
+        if is_holocene {
+            match decode_holocene_1559_params(parent.extra_data.clone()) {
+                Ok((denominator, elasticity)) => {
+                    if elasticity == 0 && denominator == 0 {
+                        return U256::from(
+                            parent
+                                .next_block_base_fee(self.base_fee_params_at_timestamp(timestamp))
+                                .unwrap_or_default(),
+                        );
+                    }
+                    let base_fee_params =
+                        BaseFeeParams::new(denominator as u128, elasticity as u128);
+                    return U256::from(
+                        parent.next_block_base_fee(base_fee_params).unwrap_or_default(),
+                    )
+                }
+                Err(e) => {
+                    panic!("Error occurred: {}", e)
+                }
+            }
+        } else {
+            U256::from(
+                parent
+                    .next_block_base_fee(self.base_fee_params_at_timestamp(timestamp))
+                    .unwrap_or_default(),
+            )
+        }
+    }
+}
+
+/// Extracts the Holcene 1599 parameters from the encoded form:
+/// <https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/holocene/exec-engine.md#eip1559params-encoding>
+pub fn decode_holocene_1559_params(extra_data: Bytes) -> Result<(u32, u32), String> {
+    if extra_data.len() < 9 {
+        return Err("Insufficient data".to_string());
+    }
+    let denominator: [u8; 4] =
+        extra_data[1..5].try_into().map_err(|_| "Failed to extract denominator".to_string())?;
+    let elasticity: [u8; 4] =
+        extra_data[5..9].try_into().map_err(|_| "Failed to extract elasticity".to_string())?;
+    Ok((u32::from_be_bytes(denominator), u32::from_be_bytes(elasticity)))
+}
```

Comment:
Let's not use a `String` error type, and instead create an error enum that represents the possible error cases